### DISCUSSION TITLE
Umlaut Problem

### DISCUSSION THREAD

**Discussion by R101G** [rern/rAudio#788] [Fixed Issues]
Is there a way to get accented characters to appear correctly? Any track name with umlauts or accents displays a question mark in place of the character.

Audacious on a desktop PC shows the characters correctly, so there is nothing wrong with the streams.

```
rAudio:~  #  mpc
 /SwissClassic_Room2.aac: Maurice Ravel - Bol?ro 
[playing] #1/1   0:23/0:00 (0%)
volume: n/a   repeat: off   random: off   single: off   consume: off

```
Could it be due to a missing font?

**Comment by klassikmann** [comment]
@R101G 
It might be a problem of your character encoding. I'm using UNICODE all the way and I have no problem with special characters. For example the character "é" is called "'LATIN SMALL LETTER E WITH ACUTE' (U+00E9)" in Unicode. The encoding in UTF-8 is 0xC3A9  (two bytes). Perhaps you are using (legacy) ISO 8859-15. Then "é" is coded as 0xE9 (one byte). 

My advice: Make a single test (recode) concerning  "Boléro".

**Comment by R101G** [comment]
@klassikmann 
Thanks for the reply. Does this involve adding a new package, or is it a matter of reconfiguring something in the existing image?

**Reply by klassikmann** [reply]
@R101G 

> Does this involve adding a new package, or is it a matter of reconfiguring something in the existing image?

No and No. I think  it's a matter of the encoding of the tag-values and/or the OS file names.
How do you generate the tag values?

**Reply by klassikmann** [reply]
@R101G 
what is mp3tag saying about your tags?

**Comment by R101G** [comment]
Sorry @klassikmann, I should have made it clear that this is a radio stream which displays OK on another platform.

I think you are right about UTF-8, though. I now have the locale command displaying  LANG=en_GB.UTF-8 instead of "C" but this has not solved the problem.

**Reply by klassikmann** [reply]
> I should have made it clear that this is a radio stream which displays OK on another platform.

o.k. different situation. I'm using different web radio stations, some of them produce correct special characters, some others don't. But what they do is beyond our control.

**Comment by R101G** [comment]
@klassikmann 

I don't see how this can be any fault of the stream providers. If Audacious can display the data correctly, so should MPD. 

There is no option to specify a character set for streams in mpd.conf, but there is for files. I wonder if this is what is missing? Unfortunately, I don't have the knowledge to go any further.

**Comment by R101G** [comment]
After more reading, I see MPD assumes UTF-8 unless told otherwise. Adding #charset=8859-15 to the url in the webradio list looks to be a simple workaround. 

How do other players detect and transcode the characters automatically, is my next question.

**Comment by klassikmann** [comment]
I used ffmpeg to make a recording of  "Radio Swiss Classic DE".  The metadata are:

>   Metadata:
    icy-br          : 96
    icy-description : SwissClassic_AAC@96Kbps
    icy-genre       : Entertainment
    icy-mimetype    : audio/aac
    icy-name        : /SwissClassic_Room2.aac
    icy-pub         : 0
    icy-url         : www.srgssr.ch
    StreamTitle     : Fr�d�ric Chopin - Mazurka h-moll op.33 Nr.4
    encoder         : Lavf59.16.100

The "�" are coded as 0xEFBFBD which is the UTF-8 coded "Replacement Character" (U+FFFD) used to replace an incoming character whose value is unknown or unrepresentable in Unicode.  As you can see the mimetype of the stream is audio/aac. In rAudio the FFmpeg decoder plugin is responsible for AAC. Therefore your question is not related directly to rAudio but to the FFmpeg plugin.

Perhaps if time allows I'll have a look into the verbose logging of MPD.

**Comment by R101G** [comment]
@klassikmann 
Thanks for your interest.

I tried disabling ffmpeg and faad in turn (and then disabled both to check there were no other decoders), and they both have the same problem. The only way I can get the correct characters is to add #charset=8859-1 to the mpc add command. This loses the default coverart image, though.

**Comment by klassikmann** [comment]
@R101G 
@rern 
rern made a PHP-script available to extract metadata from a stream, please look at https://github.com/rern/rAudio-addons/blob/main/webradio/metadata.php. Applying the script to "Radio Swiss Classic DE" (http://stream.srg-ssr.ch/m/rsc_de/aacp_96) I got
`Franz Schubert - Ouvert▒re im italienischen Stil D-Dur D 590`. Hex-value of "▒" is 0xFC, which is "ü" in ISO 8859-15. This confirms my observation using ffmpeg.

**Comment by R101G** [comment]
@klassikmann 
@rern 
Does rAudio use python-chardet, or does MPD have its own methods? Either way, it's beyond my understanding.

**Comment by rern** [comment]
`charset` setting of Icecast servers might be the culprit:
- [UTF-8 metadata in Icecast MP3 streams](https://libretime.org/docs/icecast)
- [Icecast - Mount Specific Settings](https://icecast.org/docs/icecast-2.4.1/config-file.html#mountsettings)

**Reply by klassikmann** [reply]
> `charset` setting of Icecast servers might be the culprit:

That's my assumption also.

**Comment by R101G** [comment]
@rern 
I don't understand how to use this information. Does it mean there is a way to add a charset option to a webradio entry which will not invalidate the coverart file?

EDIT: How about matching the /webradiosimg entries with the station name, rather than the url? Then #charset= could be appended to the url.

**Reply by rern** [reply]
One of these should do:
- Icecast server - Change default `charset` for all stream types to UTF-8
- WebRadio stations - Properly set `charset` on their servers to UTF-8
- MPD - Detect and convert any `charset`s to UTF-8 ( ?: [ICY metadata charset](https://github.com/MusicPlayerDaemon/MPD/issues/1002) )
- rAudio
    - Fix station coverarts when `#charset=...` appended
    - Detect and convert any `charset`s to UTF-8

**Reply by klassikmann** [reply]
> One of these should do:
> 
>     * Icecast server - Change default `charset` for all stream types to UTF-8
I assume important reasons for Icecast's decision:
`Default for non-Ogg content is now Latin-1 (aka ISO-8859-1). Ogg content still uses UTF-8.`
Verdict: Don't expect Icecast changing their default.
>     * WebRadio stations - Properly set `charset` on their servers to UTF-8
Looks promising at first sight but SRG SSR will face complaints by many users with different client software.
Verdict: Don't expect WebRadio stations to change their behaviour at least in the short-run.
>     * MPD - Detect and convert any `charset`s to UTF-8 ( ?: [ICY metadata charset](https://github.com/MusicPlayerDaemon/MPD/issues/1002) )
As the maintainer of MPD (Max Kellermann) wrote:

`Too bad Icecast doesn't allow clients to choose a charset.
No, MPD will not attempt to guess the charset, because that's error prone.`
Verdict:  Don't expect an implementation in MPD.
>     * rAudio
>       
>       * Fix station coverarts when `#charset=...` appended
Verdict: Only remaining alternative.
>       * Detect and convert any `charset`s to UTF-8
Same "error prone" procedure as above.
Verdict: Don't do it.

**Comment by R101G** [comment]
@rern 
Thanks for investigating this.

If I use the command `mpc add http://stream.srg-ssr.ch/m/rsc_de/aacp_96#charset=8859-1` the metadata displays correctly. The # symbol invalidates the image files in the webradiosimg folder, though. Appending #charset=8859-1 to the url field in the edit window works the same.

The ideal fix would be to add a charset field to the Add Webradio window, with UTF-8 as default. This would mean every station could have a different charset, if needed, and the webradiosimg file names would not need to change. 

I realise this has now become a feature request :)

**Comment by rern** [comment]
@R101G, 
@klassikmann,

Please try the UPDATE branch

**Reply by klassikmann** [reply]
Updating the existing webradio http://stream.srg-ssr.ch/m/rsc_de/aacp_96 was not successful*. The encoding parameter was not saved. I had to delete the webradio and re-create it with the encoding.

*EDIT: Changing the encoding (from 8859-1 to 8859-15) wasn't successful too

**Reply by rern** [reply]
The `#charset=...` is saved within the WebRadio file, not the URL, and used behind the scene, hidden from UI.
Play it and see the encoded characters. (`8859-1` is good for the station.)

**Reply by klassikmann** [reply]
> The `#charset=...` is saved within the WebRadio file, not the URL, and used behind the scene, hidden from UI. Play it and see the encoded characters. (`8859-1` is good for the station.)

1. I know the charset is part of the corresponding webradio-file.
2. Creation of a new webradio with non-default charset is possible and effective, but not updating an existing webradio (at least not in my environment), from the UI it looks successful, but the webradio-file is not updated.
3. I'm aware that 8859-1 is sufficient because I don't expect that the EURO-sign ("€") will appear in a music title not too often. I used 8859-11 for update testing purposes.

**Comment by R101G** [comment]
@rern 

Thanks! It seems to be working perfectly.

**Comment by R101G** [comment]
@klassikmann 
@rern 
I just added 8859-1 in the new edit window, and restarted the stream. One suggestion - if the info icon did` iconv -l `it would show a list which is easier to copy and paste than the webpage is.
